url47fcfc0133_baliosian_tau_fst_tgz-pJ8-uy_edu_fing_mina_fsa_test_old_TestComposition1J8/src/uy/edu/fing/mina/fsa/tffst/Tffst.java

```diff
@@ -0,0 +1,1712 @@
+                  State origin = inTrans.getFrom();
+                  if (origin != null) {
+                    if (!state.equals(origin)) {
+                      TfString lbout = (inTrans.labelOut.clone());
+                      lbout.addAll(t.labelOut);
+                      new Transition(origin, inTrans.labelIn.clone(),lbout , t.getTo());
+                      workingStates.add(origin);
+                    } else {
+                      if (!inTrans.labelIn.isEpsilon()) {
+                        State n = new State();
+                        new Transition(n, inTrans.labelIn, inTrans.labelOut, n);
+                        workingStates.add(n);
+                        for (Transition arriTransition2 : state.getArrivingTransitions()) {
+                          State origin2 = arriTransition2.getFrom();
+                          if (!state.equals(origin2)) {
+                            new Transition(origin2, arriTransition2.labelIn, inTrans.labelOut, n);
+                            workingStates.add(origin2);
+                          }
+                        }
+                        new Transition(n, inTrans.labelIn, t.labelOut, t.getTo());
+                      }
+                    }
+                  }
+                }
+              } else {
+                for (Transition tNext : t.getTo().getTransitions()) {
+                  if (!(tNext.labelIn.isEpsilon() && tNext.getTo().equals(tNext.getFrom()))) {
+                    TfString lbout = tNext.labelOut.clone();
+                    for (int i = 0; i < t.labelOut.size(); i++)
+                      lbout.add(i, t.labelOut.get(i));
+                    Transition tNextClone = new Transition(null, tNext.labelIn.clone(), lbout, tNext.getTo());
+                    toAdd.add(tNextClone);
+                    removeState = false;
+                  }
+                }
+              }
+            }
+          }
+        }
+        state.removeAllTransitions(toRemove);
+        state.addAllTransitions(toAdd);
+        if (removeState) workingStates.remove(state);
+      }
+    } catch (CloneNotSupportedException e) {
+      e.printStackTrace();
+    }
+    
+    removeDeadTransitions();
+  }
+
+  
+  
+  /**
+   * returns a Tffsr representing this tffst. it expands subEpochs in simpe
+   * pairts of TFs adding epsilons
+   * 
+   * @return
+   */
+  public Tffsr firstProjection() {
+   
+    Tffsr outTffsr = new Tffsr();
+    Map<State, uy.edu.fing.mina.fsa.tffsr.State> stateMap = new HashMap<State, uy.edu.fing.mina.fsa.tffsr.State>();
+
+    for (State fstState : getStates()) {
+      uy.edu.fing.mina.fsa.tffsr.State fsrState = new uy.edu.fing.mina.fsa.tffsr.State();
+      fsrState.setAccept(fstState.isAccept());
+      if (initial.equals(fstState)) 
+        outTffsr.initial = fsrState;
+      stateMap.put(fstState, fsrState);
+    }
+    
+    for (State fstState : stateMap.keySet()) {
+      uy.edu.fing.mina.fsa.tffsr.State fsrState = stateMap.get(fstState);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.